### ISSUE TITLE
ASGI/Asyncio Support

### ISSUE THREAD

**Issue by dusty-phillips** [launchdarkly/python-server-sdk#158]
**Is your feature request related to a problem? Please describe.**

As far as I can tell, this SDK won't work with asyncio or asgi services due to the use of threading locks and queue.queue. This is preventing us from using it across all of our newer services.

**Describe the solution you'd like**

I'd like to see an async-first implementation where the launch darkly task is running in the background on the same event loop.

**Describe alternatives you've considered**

I'd also consider using the existing threaded mechanism, but provide async functions that don't rely on locks and the synchronous queue to request features when necessary.

I've considered switching to a different feature flag provider.

**Additional context**

Is it possible I'm missing a different library? I was shocked to see that there was no asyncio first class citizen in the documentation in 2021, but maybe it hasn't been publicized yet.

**Comment by eli-darkly** [comment]
Async support has been requested before and it is certainly something we may add in the future, but as you can probably imagine, that would amount to a total rewrite of the SDK due to how concurrency is used throughout the code. The Python SDK codebase is older than many of the other SDKs, and was mostly written before async was available. Unfortunately I can't provide a time frame for this.

**Comment by JBogossianKH** [comment]
Any update on this? ASGI and Async support is becoming ubiquitous at this point. My team is considering using LD but ASGI support is a hard requirement as we are using FastAPI.

**Comment by prussyuval** [comment]
Any updates?

**Comment by eli-darkly** [comment]
@prussyuval No, there aren't any updates. I promise that we will post a comment here on this issue as soon as there is anything to say; that's the whole reason why we've left the issue open, so people can subscribe to it.

**Comment by rememberlenny** [comment]
Any update?

**Comment by tom-hpe-hancock** [comment]
Yes, very interested in this too for python.

**Comment by mabreuortega** [comment]
we are stuck too :-(

**Comment by Juanito98** [comment]
Any updates?

**Comment by dor1202** [comment]
@eli-darkly @eplusminus @keelerm84 No update for more than a year :worried:. Do you guys have any updates?

**Comment by rolyv** [comment]
I'm wondering if this is not a huge concern for async python based on how launchdarkly is architected. As I understand based on the LD docs, evaluating a flag should not be a blocking I/O call. Instead, the client has a local copy of the flag values and evaluation is a very fast operation. If so, it shouldn't be an issue for async code, right? I'm not an expert here, so would appreciate if any of the maintainers could comment on this.

**Comment by nikitagromov** [comment]
@rolyv They are doing sync requests in separate threads for streaming part of API

**Comment by adambirds** [comment]
This really needs to be a priority now. Hardly writing any synchronous code anymore. Even Django is heading mostly to async if you want support for websockets or channels. Any bot related stuff is all based on asyncio. This should be a priority.

**Comment by jcunhafonte** [comment]
Do you have updates?

**Comment by gabrielmbmb** [comment]
This would be really useful!

**Comment by aviadr1** [comment]
Some measurements from running this SDK in an asyncio fleet (~15 services, SDK 9.16.0), because I think this thread is asking for more than it needs to.

**@rolyv was half right** ([comment](https://github.com/launchdarkly/python-server-sdk/issues/158#issuecomment-1929318312)). Flag evaluation really is fine. It is an in-memory lookup, no I/O, no lock, and it does not block the loop. We checked.

The problem is not evaluation. It is the **lifecycle calls**, and every app makes those.

| Call | Blocks the event loop? | Measured, LaunchDarkly unreachable |
|---|---|---|
| `variation()` — the hot path | no | in-memory, no I/O |
| `LDClient(config)` | **yes** | **5.03s** (default `start_wait=5`) |
| `set_config()` | **yes** | **5.00s** |
| `close()` | **yes** | seconds, and unbounded if DNS hangs (#493) |

What a 5 second block looks like from inside the process:

```
 t=0s   ldclient.set_config(...) ───────────────────► returns at t=5s
        │
        └─ loop frozen. no HTTP handlers, no Kafka polls, no timers.

 heartbeat gaps:  0.02  0.02  0.02  [ 5.00 ]  0.02  0.02
                                       ▲
                              nothing ran for 5 seconds
```

Repro is easy: point `base_uri`/`stream_uri`/`events_uri` at `http://localhost:1` (connection refused is what an outage looks like to the SDK), call it from a coroutine, and sample the loop with a heartbeat coroutine.

This is not a rare path. `update_processor_ready.wait(start_wait)` runs to completion whenever LaunchDarkly is unreachable, so during an outage the full 5 seconds is the **normal** case, not the worst case.

**Why this matters for prioritisation**

The reason given for not doing this is a full async rewrite. Based on the table above, we did not need one. We got a fully non-blocking integration on the current SDK with two changes:

```
                          worst event-loop stall
before                            5.03s
after                             0.03s      (healthy loop is ~0.02s)
```

1. construct with `start_wait=0` and poll `is_initialized()` from a coroutine ourselves
2. run `close()` on a daemon thread with a timeout

That is it. No rewrite, no async client.

So the ask could be much smaller than "async-first". Something like:

- a documented "do not block" construction mode (`start_wait=0` already works, it is just not presented as the async story), and
- either async variants of the three lifecycle calls, or a clear note that they must be run off-loop

The background threads never bothered us. The seconds-long synchronous calls did.

Worth noting these same synchronous-call-under-a-lock shapes are behind three separate bugs: #495, #496, #497, plus the hang in #493. Making the lifecycle non-blocking would take pressure off all of them.

**Comment by jsonbailey** [comment]
We have released version [9.17.0](https://github.com/launchdarkly/python-server-sdk/releases/tag/9.17.0) which has experimental async support and are looking for feedback using the SDK. Please let us know if you experience any issues. We plan to remove the experimental label once we have received feedback from customers who wish to test the new client. It should have full compatibility with the current sync SDK with the exception on Consul persistent store support which was not included.
